File: examples/download-grc-g6a.py
#!/usr/bin/env python3
"""
Chicken Genome (GRCg6a) Downloader

Downloads CDS, FASTA, GFF, and protein files from NCBI

Usage:
  python examples/download-grc-g6a.py [outputDir] [email] [apiKey]

Example:
  python examples/download-grc-g6a.py ./chicken-genome your@email.com YOUR_API_KEY
"""

import argparse
import os
import subprocess
import time
import urllib.parse
import urllib.request
import json
import sys

BASE_URL = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils'


def delay(ms):
    """Simple delay function."""
    time.sleep(ms / 1000.0)


class GRCDownloader:
    def __init__(self, email='', api_key=''):
        self.email = email
        self.api_key = api_key
        # Rate limit: 3 req/s without key, 10 req/s with key
        self.rate_limit = 335 if not api_key else 100

    def _build_params(self, params):
        """Add email and API key to params if provided."""
        if self.email:
            params['email'] = self.email
        if self.api_key:
            params['api_key'] = self.api_key
        return params

    def search_assembly(self, assembly_name):
        """Search for assembly by name."""
        print(f"🔍 Searching for assembly: {assembly_name}...")
        
        params = self._build_params({
            'db': 'assembly',
            'term': f'{assembly_name}[Assembly Name]',
            'retmode': 'json',
            'retmax': 5
        })
        
        url = f"{BASE_URL}/esearch.fcgi?{urllib.parse.urlencode(params)}"
        
        delay(self.rate_limit)
        with urllib.request.urlopen(url) as response:
            data = json.loads(response.read().decode())
        
        ids = data.get('esearchresult', {}).get('idlist', [])
        if not ids:
            raise ValueError(f"Assembly not found: {assembly_name}")
        
        print(f"✅ Found assembly IDs: {', '.join(ids)}")
        return ids[0]

    def get_assembly_summary(self, assembly_id):
        """Get assembly summary."""
        print(f"📋 Getting assembly summary for ID: {assembly_id}...")
        
        params = self._build_params({
            'db': 'assembly',
            'id': assembly_id,
            'retmode': 'json'
        })
        
        url = f"{BASE_URL}/esummary.fcgi?{urllib.parse.urlencode(params)}"
        
        delay(self.rate_limit)
        with urllib.request.urlopen(url) as response:
            data = json.loads(response.read().decode())
        
        summary = data.get('result', {}).get(assembly_id)
        if not summary:
            raise ValueError('No summary found for assembly')
        
        ftp_path = summary.get('ftppath_genbank') or summary.get('ftppath')
        https_path = ftp_path.replace('ftp://', 'https://') if ftp_path else None
        
        print(f"📁 FTP path: {ftp_path}")
        print(f"🔗 HTTPS path: {https_path}")
        
        return summary

    def find_assembly_accession(self, assembly_name):
        """Find assembly accession and folder name."""
        # First search for the assembly
        assembly_id = self.search_assembly(assembly_name)
        
        # Get summary to find the GenBank FTP path
        summary = self.get_assembly_summary(assembly_id)
        
        # Extract FTP path
        ftp_path = summary.get('ftppath_genbank') or summary.get('ftppath')
        https_path = ftp_path.replace('ftp://', 'https://') if ftp_path else None
        
        # Extract accession (e.g., GCA_000002315.5 from /genomes/all/GCA/000/002/315/GCA_000002315.5_GRCg6a)
        accession = None
        if ftp_path:
            import re
            match = re.search(r'(GCA_\d+\.\d+)', ftp_path)
            if match:
                accession = match.group(1)
        
        print(f"🔑 GenBank Accession: {accession}")
        
        # Extract the full assembly folder name (e.g., GCA_000002315.5_GRCg6a)
        assembly_folder = ftp_path.split('/')[-1] if ftp_path else None
        print(f"📁 Assembly folder: {assembly_folder}")
        
        return {
            'assembly_id': assembly_id,
            'accession': accession,
            'assembly_folder': assembly_folder,
            'ftp_path': ftp_path,
            'https_path': https_path,
            'title': summary.get('title', '')
        }

    def download_file(self, url, output_path, max_attempts=3):
        """Download file using curl with retry logic."""
        base_delay = 5000  # 5 seconds
        
        print(f"⬇️  Downloading: {output_path}...")
        
        for attempt in range(1, max_attempts + 1):
            try:
                # Use curl for FTP downloads - more reliable than Python requests for large files
                cmd = [
                    'curl', '-#', '-L', '-o', output_path,
                    '--retry', '3',
                    '--retry-delay', '5',
                    url
                ]
                subprocess.run(cmd, check=True, capture_output=True)
                
                print(f"✅ Saved: {output_path}")
                return True
            except subprocess.CalledProcessError as error:
                is_last_attempt = attempt == max_attempts
                wait_time = base_delay * (2 ** (attempt - 1))  # Exponential backoff
                
                if is_last_attempt:
                    print(f"❌ Failed to download {output_path} after {max_attempts} attempts")
                    raise error
                
                print(f"⚠️  Attempt {attempt}/{max_attempts} failed")
                print(f"   Retrying in {wait_time/1000} seconds...")
                time.sleep(wait_time / 1000.0)

    def download_grcg6a(self, output_dir='./grc-g6a'):
        """Download GRCg6a genome files."""
        print("🧬 Starting GRCg6a Download...\n")
        
        # Ensure output directory exists
        os.makedirs(output_dir, exist_ok=True)
        print(f"📂 Created directory: {output_dir}")
        
        # Find GRCg6a assembly
        assembly_info = self.find_assembly_accession('GRCg6a')
        
        print(f"\n📋 Assembly: {assembly_info['title']}")
        print(f"🔑 Accession: {assembly_info['accession']}")
        
        ftp_base = assembly_info['ftp_path']
        accession = assembly_info['accession']
        assembly_folder = assembly_info['assembly_folder']
        
        if not ftp_base or not accession or not assembly_folder:
            raise ValueError('Could not get FTP path, accession, or assembly folder from NCBI')

        # Download files
        print('\n📥 Downloading files...\n')
        
        # Build file list - use assemblyFolder (e.g., GCA_000002315.5_GRCg6a) not just accession
        files = [
            # Genomic FASTA
            {'name': 'genomic.fna.gz', 'url': f"{ftp_base}/{assembly_folder}_genomic.fna.gz"},
            # GFF3 annotations
            {'name': 'genomic.gff.gz', 'url': f"{ftp_base}/{assembly_folder}_genomic.gff.gz"},
            # CDS sequences
            {'name': 'cds_from_genomic.fna.gz', 'url': f"{ftp_base}/{assembly_folder}_cds_from_genomic.fna.gz"},
            # Protein sequences
            {'name': 'protein.faa.gz', 'url': f"{ftp_base}/{assembly_folder}_protein.faa.gz"},
            # RNA
            {'name': 'rna.fna.gz', 'url': f"{ftp_base}/{assembly_folder}_rna.fna.gz"},
            # Gene annotations
            {'name': 'gene.gff.gz', 'url': f"{ftp_base}/{assembly_folder}_gene.gff.gz"}
        ]
        
        downloaded_files = []
        
        for file in files:
            output_path = os.path.join(output_dir, file['name'])
            try:
                self.download_file(file['url'], output_path)
                downloaded_files.append(file['name'])
            except Exception as e:
                print(f"⚠️  Skipping {file['name']} due to error: {e}")
        
        print('\n✅ Downloads complete!')
        print(f"📂 Files saved to: {output_dir}/")
        
        return {
            'assembly_info': assembly_info,
            'output_dir': output_dir,
            'files': downloaded_files
        }


def main():
    """CLI entry point."""
    parser = argparse.ArgumentParser(
        description='Download Chicken Genome (GRCg6a) files from NCBI'
    )
    parser.add_argument(
        'output_dir',
        nargs='?',
        default='./grc-g6a',
        help='Output directory (default: ./grc-g6a)'
    )
    parser.add_argument(
        'email',
        nargs='?',
        default='',
        help='Email for NCBI (optional, but recommended)'
    )
    parser.add_argument(
        'api_key',
        nargs='?',
        default='',
        help='NCBI API key (optional, for faster downloads)'
    )
    
    args = parser.parse_args()
    
    print('========================================')
    print('   Chicken Genome (GRCg6a) Downloader')
    print('========================================\n')
    
    if not args.email:
        print('💡 Tip: Add email and API key for faster downloads:')
        print('   python examples/download-grc-g6a.py ./chicken your@email.com YOUR_API_KEY\n')
    
    downloader = GRCDownloader(email=args.email, api_key=args.api_key)
    
    try:
        downloader.download_grcg6a(args.output_dir)
    except Exception as e:
        print(f'❌ Error: {e}')
        sys.exit(1)


if __name__ == '__main__':
    main()

File: examples/download-grc-g6a.js
/**
 * Chicken Genome (GRCg6a) Downloader
 * 
 * Downloads CDS, FASTA, GFF, and protein files from NCBI
 * 
 * Usage:
 *   node examples/download-grc-g6a.js [outputDir] [email] [apiKey]
 * 
 * Example:
 *   node examples/download-grc-g6a.js ./chicken-genome your@email.com YOUR_API_KEY
 */

import axios from 'axios';
import { createWriteStream, mkdirSync, existsSync } from 'node:fs';
import { pipeline } from 'node:stream/promises';
import { exec } from 'node:child_process';
import { promisify } from 'node:util';

const execAsync = promisify(exec);

const BASE_URL = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils';

// Rate limiter
const delay = (ms) => new Promise(resolve => setTimeout(resolve, ms));

class GRCDownloader {
  constructor(options = {}) {
    this.email = options.email || '';
    this.apiKey = options.apiKey || '';
    this.rateLimit = this.apiKey ? 100 : 335; // ms between requests (3 req/s without key)
  }

  async searchAssembly(assemblyName) {
    console.log(`🔍 Searching for assembly: ${assemblyName}...`);
    
    const url = `${BASE_URL}/esearch.fcgi`;
    const params = {
      db: 'assembly',
      term: `${assemblyName}[Assembly Name]`,
      retmode: 'json',
      retmax: 5
    };
    
    if (this.email) params.email = this.email;
    if (this.apiKey) params.api_key = this.apiKey;

    await delay(this.rateLimit);
    const response = await axios.get(url, { params });
    
    const ids = response.data.esearchresult?.idlist || [];
    if (ids.length === 0) {
      throw new Error(`Assembly not found: ${assemblyName}`);
    }
    
    console.log(`✅ Found assembly IDs: ${ids.join(', ')}`);
    return ids[0];
  }

  async getAssemblySummary(assemblyId) {
    console.log(`📋 Getting assembly summary for ID: ${assemblyId}...`);
    
    const url = `${BASE_URL}/esummary.fcgi`;
    const params = {
      db: 'assembly',
      id: assemblyId,
      retmode: 'json'
    };
    
    if (this.email) params.email = this.email;
    if (this.apiKey) params.api_key = this.apiKey;

    await delay(this.rateLimit);
    const response = await axios.get(url, { params });
    
    const summary = response.data.result?.[assemblyId];
    if (!summary) {
      throw new Error('No summary found for assembly');
    }
    
    // Extract FTP path and convert to HTTPS
    const ftpPath = summary.ftppath_genbank || summary.ftppath;
    console.log(`📁 FTP path: ${ftpPath}`);
    
    // Convert FTP to HTTPS URL (NCBI supports both)
    const httpsPath = ftpPath ? ftpPath.replace('ftp://', 'https://') : null;
    console.log(`🔗 HTTPS path: ${httpsPath}`);
    
    return summary;
  }

  async findAssemblyAccession(assemblyName) {
    // First search for the assembly
    const assemblyId = await this.searchAssembly(assemblyName);
    
    // Get summary to find the GenBank FTP path
    const summary = await this.getAssemblySummary(assemblyId);
    
    // Extract FTP path and convert to HTTPS
    const ftpPath = summary.ftppath_genbank || summary.ftppath;
    const httpsPath = ftpPath ? ftpPath.replace('ftp://', 'https://') : null;
    
    // Extract accession from FTP path (e.g., GCA_000002315.5 from /genomes/all/GCA/000/002/315/GCA_000002315.5_GRCg6a)
    let accession = null;
    if (ftpPath) {
      const match = ftpPath.match(/GCA_\d+\.\d+/);
      if (match) {
        accession = match[0];
      }
    }
    
    console.log(`🔑 GenBank Accession: ${accession}`);
    
    // Extract the full assembly folder name from FTP path (e.g., GCA_000002315.5_GRCg6a)
    const assemblyFolder = ftpPath ? ftpPath.split('/').pop() : null;
    console.log(`📁 Assembly folder: ${assemblyFolder}`);
    
    return {
      assemblyId,
      accession,
      assemblyFolder,
      ftpPath,
      httpsPath,
      title: summary.title
    };
  }

  async downloadFile(url, outputPath, attempt = 1, maxAttempts = 3) {
    const maxRetries = 3;
    const baseDelay = 5000; // 5 seconds
    
    console.log(`⬇️  Downloading: ${outputPath}...`);
    
    for (let attempt = 1; attempt <= maxAttempts; attempt++) {
      try {
        // Use curl for FTP downloads - more reliable than axios for large files
        const curlCmd = `curl -# -L -o "${outputPath}" --retry 3 --retry-delay 5 "${url}"`;
        await execAsync(curlCmd);
        
        console.log(`✅ Saved: ${outputPath}`);
        return true;
      } catch (error) {
        const isLastAttempt = attempt === maxAttempts;
        const delay = baseDelay * Math.pow(2, attempt - 1); // Exponential backoff
        
        if (isLastAttempt) {
          console.error(`❌ Failed to download ${outputPath} after ${maxAttempts} attempts: ${error.message}`);
          throw error;
        }
        
        console.warn(`⚠️  Attempt ${attempt}/${maxAttempts} failed: ${error.message}`);
        console.warn(`   Retrying in ${delay/1000} seconds...`);
        await delay(delay);
      }
    }
  }

  async downloadGRCg6a(outputDir = './grc-g6a') {
    console.log('🧬 Starting GRCg6a Download...\n');
    
    // Ensure output directory exists
    if (!existsSync(outputDir)) {
      mkdirSync(outputDir, { recursive: true });
      console.log(`📂 Created directory: ${outputDir}`);
    }
    
    // Find GRCg6a assembly
    const assemblyInfo = await this.findAssemblyAccession('GRCg6a');
    
    console.log(`\n📋 Assembly: ${assemblyInfo.title}`);
    console.log(`🔑 Accession: ${assemblyInfo.accession}`);
    
    const ftpBase = assemblyInfo.ftpPath;
    const accession = assemblyInfo.accession;
    const assemblyFolder = assemblyInfo.assemblyFolder;
    
    if (!ftpBase || !accession || !assemblyFolder) {
      throw new Error('Could not get FTP path, accession, or assembly folder from NCBI');
    }

    // Download files
    console.log('\n📥 Downloading files...\n');
    
    // Build file list - use assemblyFolder (e.g., GCA_000002315.5_GRCg6a) not just accession (GCA_000002315.5)
    const files = [
      // Genomic FASTA
      { name: 'genomic.fna.gz', url: `${ftpBase}/${assemblyFolder}_genomic.fna.gz` },
      // GFF3 annotations
      { name: 'genomic.gff.gz', url: `${ftpBase}/${assemblyFolder}_genomic.gff.gz` },
      // CDS sequences
      { name: 'cds_from_genomic.fna.gz', url: `${ftpBase}/${assemblyFolder}_cds_from_genomic.fna.gz` },
      // Protein sequences  
      { name: 'protein.faa.gz', url: `${ftpBase}/${assemblyFolder}_protein.faa.gz` },
      // RNA
      { name: 'rna.fna.gz', url: `${ftpBase}/${assemblyFolder}_rna.fna.gz` },
      // Gene annotations
      { name: 'gene.gff.gz', url: `${ftpBase}/${assemblyFolder}_gene.gff.gz` }
    ];

    const downloadedFiles = [];
    
    for (const file of files) {
      const outputPath = `${outputDir}/${file.name}`;
      try {
        await this.downloadFile(file.url, outputPath);
        downloadedFiles.push(file.name);
      } catch (error) {
        console.error(`⚠️  Skipping ${file.name} due to error`);
      }
    }
    
    console.log('\n✅ Downloads complete!');
    console.log(`📂 Files saved to: ${outputDir}/`);
    
    return {
      assemblyInfo,
      outputDir,
      files: downloadedFiles
    };
  }
}

// Export for use as module
export { GRCDownloader };

// CLI usage
async function main() {
  const args = process.argv.slice(2);
  const outputDir = args[0] || './grc-g6a';
  const email = args[1] || '';
  const apiKey = args[2] || '';

  console.log('========================================');
  console.log('   Chicken Genome (GRCg6a) Downloader');
  console.log('========================================\n');

  if (!email) {
    console.log('💡 Tip: Add email and API key for faster downloads:');
    console.log('   node examples/download-grc-g6a.js ./chicken your@email.com YOUR_API_KEY\n');
  }

  const downloader = new GRCDownloader({ email, apiKey });

  try {
    await downloader.downloadGRCg6a(outputDir);
  } catch (error) {
    console.error('❌ Error:', error.message);
    if (error.response) {
      console.error('📄 NCBI Response:', error.response.status, error.response.statusText);
    }
    process.exit(1);
  }
}

// Always run main when executed directly
main();
